### PR TITLE
Added warning if local worker connection fails. Added mitigation advice to Core SDK download failure.

### PR DESCRIPTION
**Pull request by callumlawson**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Added mitigation advice to failed Core Sdk download.
Added warning if worker connection fails in editor.
image

#### Tests
Ran locally both in the failing and passing case.
#### Documentation
None.

### CODE REVIEW

File: workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs [hunk 1]
```diff
@@ -54,6 +54,11 @@ public async Task Connect(string workerType, ILogDispatcher logger)
                         new LogEvent("Cancelling connection attempt")
                             .WithField("WorkerType", workerType)
                             .WithField("Reason", "Required worker failed to connect"));
+#if UNITY_EDITOR
+                    logger.HandleLog(LogType.Warning,
+                        new LogEvent("A local runtime may not have been started. If not, you can start one from 'SpaitalOS -> Local launch' or by pressing Ctrl-L")
```
Comment (samiwh) [inline workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: I'd reword this as something you have to check unless we are checking that this is the likely reason. If I read this I'd think it had failed to detect a runtime, not that it was just a guess.

Comment (jared-improbable) [inline workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: That's not accurate on Mac, where it will be "Cmd" instead of "Ctrl".

File: workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs [hunk 2]
```diff
@@ -54,6 +54,11 @@ public async Task Connect(string workerType, ILogDispatcher logger)
                         new LogEvent("Cancelling connection attempt")
                             .WithField("WorkerType", workerType)
                             .WithField("Reason", "Required worker failed to connect"));
+#if UNITY_EDITOR
+                    logger.HandleLog(LogType.Warning,
```
Comment (samiwh) [inline workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: Arguably this is actually a Debug.Log as this is a pure editor message and not a worker message.

File: workers/unity/Packages/com.improbable.gdk.tools/DownloadCoreSdk.cs [hunk 3]
```diff
@@ -120,7 +120,7 @@ private static DownloadResult Download()
                     exitCode = RedirectedProcess.Run(Common.DotNetBinary, ConstructArguments());
                     if (exitCode != 0)
                     {
-                        Debug.LogError($"Failed to download SpatialOS Core Sdk version {Common.CoreSdkVersion}.");
+                        Debug.LogError($"Failed to download SpatialOS Core Sdk version {Common.CoreSdkVersion}. You can use SpatialOS -> Download CoreSDK (force) to retry this.");
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/DownloadCoreSdk.cs]: These menu item names should probably be constants interpolated into the string, so this message stays up to date when things invariably change.

File: workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs [hunk 4]
```diff
@@ -54,6 +54,11 @@ public async Task Connect(string workerType, ILogDispatcher logger)
                         new LogEvent("Cancelling connection attempt")
                             .WithField("WorkerType", workerType)
                             .WithField("Reason", "Required worker failed to connect"));
+#if UNITY_EDITOR
```
Comment (samiwh) [inline workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: Is there a ticket for doing this properly? We've discussed several less hacky ways of doing this so it might be worth commenting on this to that effect.

File: workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs (hunk 1 above)
Comment (callumlawson) [reply workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: Ah, good point! I assume Cmd/Ctrl-L would probably be acceptable?

File: workers/unity/Packages/com.improbable.gdk.tools/DownloadCoreSdk.cs (hunk 3 above)
Comment (callumlawson) [reply workers/unity/Packages/com.improbable.gdk.tools/DownloadCoreSdk.cs]: I agree, although currently trying to be suuuuper low impact with these changes. Ill make a ticket.

File: workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs (hunk 4 above)
Comment (callumlawson) [reply workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: I'll make a ticket for this

File: workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs (hunk 2 above)
Comment (callumlawson) [reply workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: I wasn't sure, as it is a "warning" in the sense that it is warning the user about a potential issue. Do you think it is never appropriate to use a warning for a notification like that?

File: workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs (hunk 1 above)
Comment (callumlawson) [reply workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: Will do

**Comment by callumlawson** [comment]
Ticket for moving "SpaitalOS" to a shared constant string:
https://improbableio.atlassian.net/browse/UTY-1147

Ticket for detecting if a local runtime is running and using that info to alert the user:
https://improbableio.atlassian.net/browse/UTY-1148

File: workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs [hunk 5]
```diff
@@ -54,6 +54,12 @@ public async Task Connect(string workerType, ILogDispatcher logger)
...
+                    //Temporary warning to be replaced when we can reliably detect if a local runtime is running, or not.
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: Super super nit: `// Temporary..`

File: workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs [hunk 6]
```diff
@@ -54,6 +54,12 @@ public async Task Connect(string workerType, ILogDispatcher logger)
                         new LogEvent("Cancelling connection attempt")
                             .WithField("WorkerType", workerType)
                             .WithField("Reason", "Required worker failed to connect"));
+#if UNITY_EDITOR
+                    //Temporary warning to be replaced when we can reliably detect if a local runtime is running, or not. 
+                    logger.HandleLog(LogType.Warning,
+                        new LogEvent("Is a local runtime running? If not, you can start one from 'SpaitalOS -> Local launch' or by pressing Cmd/Ctrl-L")
```
Comment (jessicafalk) [inline workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs]: typo: `SpaitalOS` -> `SpatialOS`
